### ISSUE TITLE
PiranhaJava - Unable to remove stale Unit Test code 

### ISSUE THREAD

**Issue by Samhithgb** [uber/piranha#43]
Hi, 

I'm defining an annotation : 

```
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.METHOD})
public @interface FlagTesting {
    MyClass.TestExperimentName[] treated();
}

```
and using it in my unit test : 

```
@Test
@FlagTesting(treated = MyClass.TestExperimentName.SAMPLE_STALE_FLAG)
public void addition_isCorrect() throws Exception {
...
}

```
When I build this, although the flag and its corresponding methods in the product code get removed as expected, unit test build fails with the error : 

```
Execution failed for task ':app:compileInternalChromeBeta64DebugUnitTestJavaWithJavac'.
> Piranha is not a valid checker name
```

Any pointers on why this might happen will be really great. 

PS : Using this on Android code that uses gradle.

**Comment by mkr-plse** [comment]
Have you updated the piranha.properties file?

**Comment by Samhithgb** [comment]
@mkr-plse Yes, this is how it looks : 

> treatedMethods=IsFeatureEnabled,flagEnabled
> controlMethods=flagDisabled
> emptyMethods=enableFlag,disableFlag
> annotations=FlagTesting
> linkURL=<provide_your_url>
> 

The methods listed get removed out as expected.

**Comment by lazaroclapp** [comment]
This sounds like Piranha itself is not being loaded (but Error Prone is and is configured to run piranha). Do you have a sample gradle file showing the issue? 

`:app:compileInternalChromeBeta64DebugUnitTestJavaWithJavac` sounds like it potentially uses a custom `javac` command line. Make sure Piranha is on the classpath for it.

**Comment by Samhithgb** [comment]
@lazaroclapp Thanks for you response. 

Adding Piranha to errorprone configuration resolved the issue seems like.  

However, there is a new issue : The stale flag gets removed from the product code initially and then the unit test build task fails comlaining : 

> /Users/samhith/RfAndroid/UI/app/src/test/java/com/citrix/client/Receiver/ExampleUnitTest.java:28: error: cannot find symbol
> @FlagTesting(treated = MyClass.TestExperimentName.SAMPLE_STALE_FLAG)
> 

My gradle file currently looks something like this : 

```

plugins {
    id "com.github.sherter.google-java-format" version "0.7.1"
    id "net.ltgt.errorprone" version "0.6"
}

...
android {
    ...
    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8

    }
    ...
}

tasks.withType(Test) {
    testLogging {
        exceptionFormat 'full'
        showCauses true
        showExceptions true
        showStackTraces true
        showStandardStreams true
    }
}

dependencies {
...
    annotationProcessor "com.uber.piranha:piranha:0.0.3"
    errorprone "com.uber.piranha:piranha:0.0.3"
    errorprone "com.google.errorprone:error_prone_core:2.3.2"
    errorproneJavac "com.google.errorprone:javac:9+181-r4173-1"
...
}


afterEvaluate {
    tasks.withType(JavaCompile) {
        options.errorprone.allErrorsAsWarnings = true
        options.errorprone {
            check("Piranha", net.ltgt.gradle.errorprone.CheckSeverity.WARN)
        }
        options.errorprone.errorproneArgs << "-XepPatchChecks:Piranha"
        options.errorprone.errorproneArgs << "-XepPatchLocation:IN_PLACE"
        // The lines below should be replaced by code that loads the specific flag to patch
        // and final treatment condition.
        options.errorprone.errorproneArgs << "-XepOpt:Piranha:FlagName=SAMPLE_STALE_FLAG"
        options.errorprone.errorproneArgs << "-XepOpt:Piranha:IsTreated=true"
        options.errorprone.errorproneArgs << "-XepOpt:Piranha:Config=config/piranha.properties"
    }
}
```

**Comment by mkr-plse** [comment]
Can you run it on the test code before the production code? What is happening is that the flag definition gets removed from production and the compilation fails in the test.

**Comment by Samhithgb** [comment]
Thanks for a quick respnse. 

Yes, but the corresponding boolean flag methods inside the unit test are getting removed though. 

For example, 

  ```
@Test
    @FlagTesting(treated = MyClass.TestExperimentName.SAMPLE_STALE_FLAG)
    public void addition_isCorrect() throws Exception {
        if (MyClass.XPTest.flagEnabled( MyClass.TestExperimentName.SAMPLE_STALE_FLAG, "Testing")) {
            System.out.println("Hello World");
        } else {
            System.out.println("Hello World again");
        }
    }
```

is getting modified into : 


```
 @Test
    @FlagTesting(treated = MyClass.TestExperimentName.SAMPLE_STALE_FLAG)
    public void addition_isCorrect() throws Exception {
        System.out.println("Hello World");
    }
```

Shouldn't the annotation be removed as well along with methods?

**Comment by mkr-plse** [comment]
Either the annotation should be removed or the method implementation will be removed. See [here](https://github.com/uber/piranha/blob/master/java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java#L730). 

Can you try a few things: 
1) delete @Test annotation. 
2) instead of the fully qualified name for the flag in the annotation, just use the flag name

Do any of these changes affect the refactoring?

**Comment by Samhithgb** [comment]
@mkr-plse  I tried both of these, but unfortunately still have issues with removing the annotations. 

Initial : 

```
 @FlagTesting(treated = SAMPLE_STALE_FLAG)
    public void addition_isCorrect() throws Exception {
        if (MyClass.XPTest.flagEnabled( SAMPLE_STALE_FLAG, "Testing")) {
            System.out.println("Hello World");
        } else {
            System.out.println("Hello World again");
        }
    }

```

After build : 

```
 @FlagTesting(treated = SAMPLE_STALE_FLAG)
    public void addition_isCorrect() throws Exception {
        System.out.println("Hello World");
    } 
```

Is there anything else I can try?

**Comment by Samhithgb** [comment]
This is what I see in build logs : 

> > Task :app:compileNormal64ReleaseUnitTestJavaWithJavac
> /Users/samhith/RfAndroid/UI/app/src/test/java/com/citrix/client/Receiver/ExampleUnitTest.java:15: warning: [Piranha] Cleans stale XP flags.
> import static com.citrix.client.Receiver.FeatureFlag.MyClass.TestExperimentName.SAMPLE_STALE_FLAG;
> ^
>   Did you mean 'import static com.citrix.client.Receiver.util.ReceiverContext.PlatformHelper.isNetworkAvailable;'?
> /Users/samhith/RfAndroid/UI/app/src/test/java/com/citrix/client/Receiver/ExampleUnitTest.java:28: warning: [Piranha] Cleans stale XP flags.
>         if (MyClass.XPTest.flagEnabled( SAMPLE_STALE_FLAG, "Testing")) {
>         ^
>   Did you mean 'System.out.println("Hello World");'?
> 
>

**Comment by mkr-plse** [comment]
Update your `piranha.properties` file (which I assume is the one from [here](https://github.com/uber/piranha/blob/master/java/sample/config/piranha.properties)) to include `treatmentGroupMethods`. See the config file that contains all properties [here](https://github.com/uber/piranha/blob/master/java/piranha/config/piranha.properties).

The problem is that this [line](https://github.com/uber/piranha/blob/master/java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java#L273) returns a `null` for the property `treatmentGroupMethods` since it is absent and subsequently a NPE is thrown. The remaining code in `init` is not processed resulting in annotations not being handled. 

While this is a temporary fix for you, I will re-purpose this issue to fix the NPE.

**Comment by Samhithgb** [comment]
Great, that worked. Thank you! 

Just FYI : There seems to be no documentation available on treatmentGroupMethods field. What does treatmentGroupMethods do? Also, the example refactoring [here](https://github.com/uber/piranha/blob/master/java/README.md#example-refactoring) misses the field as well.

**Comment by mkr-plse** [comment]
It doesn't do anything special now and will soon be removed.

**Comment by mkr-plse** [comment]
Fixed in #45.
